bugwarrior/services/todoist.py

```diff
@@ -0,0 +1,220 @@
+
+    def get_sections(self):
+        all_sections = []
+        sections_iter = self._api.get_sections()
+        for sections in sections_iter:
+            for section in sections:
+                all_sections.append(section)
+        return all_sections
+
+    def get_users(self, project_id):
+        all_users = []
+        users_iter = self._api.get_collaborators(project_id)
+        for users in users_iter:
+            for user in users:
+                all_users.append(user)
+        return all_users
+
+    def get_issues(self):
+        tasks_iter = self._api.filter_tasks(query=self.filter)
+        return tasks_iter
+
+
+class TodoistIssue(Issue):
+    ASSIGNEE = "todoistassignee"
+    ASSIGNER = "todoistassigner"
+    CONTENT = "todoistcontent"
+    DESCRIPTION = "todoistdescription"
+    DURATION = "todoistduration"
+    ID = "todoistid"
+    SECTION = "todoistsection"
+    URL = "todoisturl"
+
+    PRIORITY_MAP = {
+        4: "H",
+        3: "M",
+        2: "L",
+        1: None,
+    }
+
+    UDAS = {
+        ID: {
+            "type": "string",
+            "label": "Todoist ID",
+        },
+        CONTENT: {
+            "type": "string",
+            "label": "Todoist Content",
+        },
+        DESCRIPTION: {
+            "type": "string",
+            "label": "Todoist Description",
+        },
+        DURATION: {
+            "type": "string",
+            "label": "Todoist Duration",
+        },
+        SECTION: {
+            "type": "string",
+            "label": "Todoist Section",
+        },
+        ASSIGNEE: {
+            "type": "string",
+            "label": "Todoist Assignee",
+        },
+        ASSIGNER: {
+            "type": "string",
+            "label": "Todoist Assigner",
+        },
+        URL: {
+            "type": "string",
+            "label": "Todoist URL",
+        },
+    }
+
+    UNIQUE_KEY = (ID, ID)
+
+    # replace characters that cause escaping issues in teh description like [] and "
+    # this is a workaround for https://github.com/ralphbean/taskw/issues/172
+    def _unescape_content(self, content):
+        return (
+            content.replace('"', "'")  # prevent &dquote; in task details
+            .replace("[", self.config.char_open_bracket)  # prevent &open; and &close;
+            .replace("]", self.config.char_close_bracket)
+        )
+
+    def to_taskwarrior(self):
+        default_time = time(0, 0, 0)
+        # use due date "scheduled".
+        # adjust timezone to use local time for "floating" dates
+        if self.record.due and type(self.record.due.date) is datetime:
+            if self.record.due.timezone:
+                scheduled = self.record.due.date
+            else:
+                scheduled = self.record.due.date.replace(tzinfo=None)
+        else:
+            scheduled = (
+                datetime.combine(self.record.due.date, default_time, tzinfo=None)
+                if self.record.due
+                else None
+            )
```

Comment:
I believe this is logically identical but easier to follow.
```suggestion
        if self.record.due:
            if type(self.record.due.date) is datetime:
                if self.record.due.timezone:
                    scheduled = self.record.due.date
                else:
                    scheduled = self.record.due.date.replace(tzinfo=None)
            else:
                scheduled = datetime.combine(self.record.due.date, default_time, tzinfo=None)
        else:
            scheduled = None
```

Reply:
Looking briefly through some Todoist docs, it seems like the way you map due->scheduled and deadline->due is the implied suggestion. However, it seems totally valid for users to use Todoist's due date as an actual due date and they might be unpleasantly surprised at this behavior if for example their tasks don't show up in a report until they're overdue. So I'm thinking this should be an optional, non-default behavior.

Reply:
I think that makes sense. Need to think through what the choices would be:
- option 1: always due→due, ignore deadline
- option 2: due→due unless deadline is set, then due→scheduled and deadline→due
- option 3: always due→scheduled and deadline→due

Reply:
Giving this some more thought, I think option 2 should be the default (use due as due unless deadline is set).

Proposed approach: new option setting `todoist.due_date_mapping`
- `default` - due to due unless deadline, then due to scheduled and deadline to due
- `always_due` - always map due to due and ignore deadline
- `always_scheduled` - always map due to scheduled and deadline to due.

Reply:
The problem I see with that approach is that people might use `due` as a "scheduled" date but not set `deadline` for every task. I think that while users will have different mapping preferences, the mapping should be consistent for all tasks.

I'd forgotten that we'd discussed this before in #385 and determined that the best course of action is generally to add such fields as due and deadline as UDA's. I don't think any other services do so but I wouldn't be opposed to documenting a suggested use of due_template/scheduled_template as a Service Feature if you wanted.

Reply:
The default behavior I added I think aligns best to the Todoist philosophy for dates vs deadlines [The difference between a date and a deadline](https://www.todoist.com/help/articles/introduction-to-deadlines-uMqbSLM6U#h_01JDS5KZG9AMRPBWBEK366TXGE), and the other options should cover user specific alternative.  I'll add the UDA's as well.

The due_template/schedule_template is an interesting idea, but probably overkill

Reply:
> The due_template/schedule_template is an interesting idea, but probably overkill

All you have to do to get this functionality is add UDA's for due and schedule thanks to [field templates](https://bugwarrior.readthedocs.io/en/latest/common_configuration.html#field-templates).

Reply:
UDA's added. Does that resolve this thread - or are any other changes needed?

Reply:
Frankly, I think the `due_date_mapping` option should be removed and the mapping in logseq should likewise be removed.

I'm not really convinced that the default behavior matches the Todoist philosophy. I'd seen that article and to me it seems intentionally ambiguous to paper over the fact that their field semantics don't make any sense. I find it hard to believe that the default behavior represents the way people use Todoist -- using `deadline` as a due date for some tasks and `due` as a due date for other tasks. I would expect the meaning of these fields to be consistent within an organization.

The other point is that bugwarrior services don't generally import dates into date fields and we have (more or less successfully) been intentional about this since #385.

Reply:
Okay, so:
- [ ] Remove the `due_date_mapping` config option
- [ ] only map Todoist due date to Taskwarrior due date. Ignore Todoist deadline dates. Nothing is mapped to taskwarrior scheduled date
- [x] Todoist Due and Deadline dates added as UDAs
- [ ] Document how to use `due_template` and `scheduled_template` to alter the mapping

The following templates work the for behavior I was suggesting:
```
todoist.due_template = {{ todoistdeadline if todoistdeadline else todoistdue if todoistdue else "" }}
todoist.scheduled_template = {{ todoistdue if todoistdeadline else "" }}
```